src/raftstore/store/worker/region.rs

```diff
@@ -305,24 +451,240 @@ impl Runnable<Task> for Runner {
                 region_id,
                 notifier,
             } => {
-                // It safe for now to handle generating and applying snapshot concurrently,
+                // It is safe for now to handle generating and applying snapshot concurrently,
                 // but it may not when merge is implemented.
                 let ctx = self.ctx.clone();
                 self.pool
                     .execute(move |_| ctx.handle_gen(region_id, notifier))
             }
-            Task::Apply { region_id, status } => self.ctx.handle_apply(region_id, status),
+            Task::Apply { region_id, status } => {
+                self.ctx
+                    .handle_apply(&mut self.pending_delete_ranges, region_id, status)
+            }
             Task::Destroy {
                 region_id,
                 start_key,
                 end_key,
-            } => self.ctx.handle_destroy(region_id, start_key, end_key),
+            } => {
+                if self.clean_stale_peer_delay.as_secs() > 0 {
+                    info!(
+                        "[region {}] register deleting data in [{}, {})",
+                        region_id,
+                        escape(&start_key),
+                        escape(&end_key)
+                    );
+                    // delay the range deletion becase there might be a coprocessor request related to this range
+                    let timeout = time::Instant::now() + self.clean_stale_peer_delay;
+                    self.pending_delete_ranges
+                        .insert(region_id, start_key, end_key, timeout);
+                } else {
+                    self.ctx.handle_destroy(region_id, start_key, end_key);
+                }
+            }
+            Task::CleanStalePeer {} => {
+                STALE_PEER_GAUGE_VEC
+                    .with_label_values(&["total"])
+                    .set(self.pending_delete_ranges.len() as f64);
+
+                let now = time::Instant::now();
+                let mut timeout_ranges = self.pending_delete_ranges.drain_timeout_ranges(now);
+                STALE_PEER_GAUGE_VEC
```

Comment:
This metric seems not useful, I think we just record the number of pending stale peers.